_posts/2021-05-03-deploying-kafka-with-lets-encrypt-certificates.md

```diff
@@ -0,0 +1,485 @@
+And you will need to update the public key when it changes, for example, because the old one expired.
+This would be much easier if the listeners can use TLS certificates signed by a certification authority which the clients trust out of the box.
+You would not need to distribute any certificates, you just tell the client to use TLS.
+And in this blog post I will show you how to do it with the help of [Let's Encrypt](https://letsencrypt.org/), [cert-manager](https://cert-manager.io/), and [ExternalDNS](https://github.com/kubernetes-sigs/external-dns).
+
+<!--more-->
+
+## Let's Encrypt
+
+[Let's Encrypt](https://letsencrypt.org/) is a non-profit certification authority which provides an automated way of issuing signed certificates which are trusted by most operating systems and software platforms. 
+It is currently providing more than 200 million certificates.
+All you need to be able to get a signed certificate from Let's Encrypt is to own a domain and be able to prove its ownership.
+Let's Encrypt is using [Automatic Certificate Management Environment (ACME) protocol](https://tools.ietf.org/html/rfc8555) to verify the domain ownership and request / issue the signed certificates.
+
+## Using the ACME protocol
+
+Don't get scared by the IETF specification of the ACME protocol.
+We will use a tool called [cert-manager](https://cert-manager.io/) to communicate with Let's Encrypt.
+All we will need to do is to configure the cert-manager and request certificates by creating custom resources in our Kubernetes cluster.
+
+When proving domain ownership, the ACME protocol supports two different mechanisms.
+One is _HTTP Challenge_ and the second is _DNS Challenge_.
+The [HTTP Challenge](https://letsencrypt.org/docs/challenge-types/#http-01-challenge) uses HTTP to verify that you own the domain.
+It requires you to put on some path on that domain some specific string which it will try to access using HTTP protocol.
+For example, when requesting a certificate for domain _example.com_ it will ask you to place a file with some specific information on the HTTP address `https://example.com/.well-known/acme-challenge/...`.
+If you are able to do it, you have proven that you control the domain.
+This works well for web applications which use HTTP, but it doesn't work so well for applications using TCP such as Apache Kafka.
+
+Apache Kafka does not use HTTP so it cannot easily provide information on a given HTTP URL.
+That is why with Apache Kafka, we will need to use the DNS Challenge.
+The [DNS Challenge](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge) works similarly, but uses the DNS protocol instead of HTTP.
+It will ask us to create a DNS record with some specific information to prove our domain ownership.
+Using the DNS challenge makes it a bit more complicated, because we will need to give cert-manager access to the DNS management of our domain.
```

Comment:
```suggestion
Using the DNS Challenge makes it a bit more complicated, because we will need to give cert-manager access to the DNS management of our domain.
```